# 스프링의 강의 리뷰📽

> LoadMap Part : 스프링 DB 1편 - 데이터 접근 핵심 원리 - 2022-05-15  
> Section : 6.스프링과 문제 해결 - 예외처리, 반복   
> CreateDate : 2022.11.07  
> UpdateDate : 2022.11.

## 사용한 스펙

> Spring Boot 버전 : Spring Boot 2.7.3  
> Java 버전 : java 11    
> Gradle 버전: gradle-7.5

<br></br>
<br></br>

# 1. 체크 예외와 인터페이스
 - 인터페이스를 통해, 구현기술을 쉽게 변경할 수 있게 해야한다. 

### 예시 코드
```java
public interface MemberRepository {
     Member save(Member member);
     Member findById(String memberId);
     void update(String memberId, int money);
     void delete(String memberId);
}
```
 - 하지만, 왜냐하면 SQLException 같은 체크 예외이기 때문에, 문제가 발생한다. 

```java
import java.sql.SQLException; // 종속적이게 됨

public interface MemberRepositoryEx {
     Member save(Member member) throws SQLException;
     Member findById(String memberId) throws SQLException;
     void update(String memberId, int money) throws SQLException;
     void delete(String memberId) throws SQLException;
}
```
 - 인터페이스의 메서드에 해당 체크 예외를 던질 수 있는 `throws SQLException` 가 있어야 한다는 것이다.
 - 즉, 순수해지지 않아진다.
> 그래서, 각 구현기술들은 런타임예외를  적용하게 된다.
> 그럼 별도로 던지는 throw처리가 없어져 특정 기술에 종속적일 필요가 없어진다는 것이다.


# 2. 런타임 예외 적용

### 예시 코드
- DB 예외 처리를 런타임 예외로 바꿔주는 코드이다.
```java

public class MyDBException extends RuntimeException{
    public MyDBException() {
    }

    public MyDBException(String message) {
        super(message);
    }

    public MyDBException(String message, Throwable cause) {
        super(message, cause);
    }

    public MyDBException(Throwable cause) {
        super(cause);
    }
}
```

- 덕분에, 위에서 못하던, 종속적이지 않는 Repository 인터페이스를 만들 수 있다.
```java
public interface MemberRepository {
     Member save(Member member);
     Member findById(String memberId);
     void update(String memberId, int money);
     void delete(String memberId);
}
```
- 그럼 이제 구현되는 repository는 위의 인터페이스를 상속 받고 DI 사용하면 예외 누수문제를 해결할 수 있다.
### 테스트 코드 
- [ 예외 누수 문제 해결, SQLException 제거한 Repository코드](/practice_code/repository/MemberRepositoryV4_1.java)
- [ 예외 누수 문제 해결, SQLException 제거한 테스트 코드](/test_code/service/MemberServiceV4Test.java)


# 3. 데이터 접근 예외 직접 만들기 
> 데이터베이스 오류에 따라서 특정 예외는 복구하고 싶을 수도 있다.   
> 예를 들어서   
> 회원 가입시 DB에 이미 같은 ID가 있으면 ID 뒤에 숫자를 붙여서 새로운 ID를 만들어야 한다고 가정해보자.
## 예시 상황 
> ID를 hello 라고 가입 시도 했는데, 이미 같은 아이디가 있으면 hello12345 와 같이 뒤에 임의의 숫자를 붙여서 가입하는 것이다.
 - 이경우, DB에 저장할때 이미 같은 ID가 있다면, 데이터베이스는 오류코드를 반환하고
   - 이 오류 코드를 받은 JDBC드라이버는 `SQLException`을 던지다.
   - 그리고 이 `SQLException`은 errorCode을 가지고 있다.
 - 이 오류코드를 활용하여, 중복된 ID가 있을 예외를 따로 처리할 수 있게 되는 것이다

<br>

### 데이터베이스 오류 코드 흐름 

<p align="center"><img src="https://user-images.githubusercontent.com/104331549/200214906-594acfd4-0dc8-46f4-9ef3-54068c5faf12.png" width="80%"></p>


- H2 데이터베이스의 경우, 키중복 오류 코드는 `23505`이다.
      - SQL 문법오류는 `42000`
      - 이처럼 오류 코드로 인해, 원하는 오류를 핸드링이 가능해진다.
      - [H2 오류코드 리스트](https://www.h2database.com/javadoc/org/h2/api/ErrorCode.html)
> 참고로 같은 오류여도 각각의 데이터베이스마다 정의된 오류 코드가 다르다.   
> **키 중복 오류 코드**  
> H2 DB: 2350  
> MySQL: 1062


<br>

### 예시 코드
- 이 문제를 해결하려면 앞서 배운 것 처럼 리포지토리에서 예외를 변환해서 던지면 된다
  - SQLException -> MyDuplicateKeyException
  - MyDBException을 상속 받는 예외클래스로, 키가 중복될 때 MyDuplicateKey로 별도로 예외 처리 할 수 있겠끔 만든 구현체이다.
- 이 예외는 우리가 직접 만든 것이기 때문에, JDBC나 JPA 같은 특정 기술에 종속적이지 않다
```java
public class MyDuplicateKeyException extends  MyDBException{
    public MyDuplicateKeyException() {
    }

    public MyDuplicateKeyException(String message) {
        super(message);
    }

    public MyDuplicateKeyException(String message, Throwable cause) {
        super(message, cause);
    }

    public MyDuplicateKeyException(Throwable cause) {
        super(cause);
    }
}
```
 - 예외를 변환하는 작업을 하는 레포지포리 테스트 코드는 아래와 같다.
   - `e.getErrorCode()`  : 에러의 오류 코드를 반환한다.
```java
public class ExTranslatorV1Test {
    // ..생략
    // 실제로, 예외를 변경하는 부분인 Repository 파트를 보면
    @RequiredArgsConstructor
    static class Repository {
        private final DataSource dataSource;

        public Member save(Member member) {
            String sql = "insert into member(member_id, money) value(?,?)";
            Connection con = null;
            PreparedStatement pstmt = null;

            try {
                con = dataSource.getConnection();
                pstmt = con.prepareStatement(sql);
                pstmt.setString(1, member.getMemberId());
                pstmt.setInt(2, member.getMoney());
                pstmt.executeUpdate();
                return member;
            } catch (SQLException e) {
                //h2 db
                if (e.getErrorCode() == 23505) { // 오류코드를 확인하여 별도 처리
                    throw new MyDuplicateKeyException(e);
                }
                throw new MyDBException(e); // 그외엔 그대로 던진다.
            } finally {
                JdbcUtils.closeStatement(pstmt);
                JdbcUtils.closeConnection(con);
            }
        }
    }
}
```
- 그리고 반환된 에러중 특정 에러만 처리하는 서비스 테스트 코드는 아래와 같다.

```java
public class ExTranslatorV1Test {
    // ..생략
    @RequiredArgsConstructor
    static class Service {
        private final Repository repository;

        public void create(String memberId) {
            try {
                repository.save(new Member(memberId, 0));
                log.info("saveId={}", memberId);
            } catch (MyDuplicateKeyException e) {
                log.info("키 중복, 복구 시도");
                String retryId = generateNewId(memberId);
                repository.save(new Member(retryId, 0));
            } catch (MyDBException e) {
                log.info("데이터 접근 계층 예외", e);
                throw e;
            }
        }

        private String generateNewId(String memberId) {
            return memberId + new Random().nextInt(1000);
        }

    }
}
```

### 참고 링크
- 테스트를 위해 코드를 입력하는 과정에서, `Syntax error in SQL statement` 에러가 발생했다. 
  - 이유는 SQL 명령어에서, value로 입력한게 이유였다. 
  - 어디에서는 value 로 사용해도 가능은 하지만, 대부분 values로 바뀌어서 values를 쓰도록 하자.
  - [insert into value, values 차이](https://stackoverflow.com/questions/17445583/what-is-difference-between-insert-value-and-insert-values-in-mysql-statement)